Content.Server/Salvage/SalvageSystem.cs

```diff
@@ -0,0 +1,344 @@
+            {
+                case MagnetStateType.Attaching:
+                    if (SpawnSalvage(magnet))
+                    {
+                        magnet.MagnetState = new MagnetState(MagnetStateType.Holding, currentTime + HoldTime);
+                    }
+                    else
+                    {
+                        magnet.MagnetState = new MagnetState(MagnetStateType.CoolingDown, currentTime + CooldownTime);
+                    }
+                    break;
+                case MagnetStateType.Holding:
+                    Report("salvage-system-announcement-losing", ("timeLeft", DetachingTime.TotalSeconds));
+                    magnet.MagnetState = new MagnetState(MagnetStateType.Detaching, currentTime + DetachingTime);
+                    break;
+                case MagnetStateType.Detaching:
+                    if (magnet.AttachedEntity.HasValue)
+                    {
+                        SafeDeleteSalvage(magnet.AttachedEntity.Value);
+                    }
+                    else
+                    {
+                        Logger.ErrorS("salvage", "Salvage detaching was expecting attached entity but it was null");
+                    }
+                    Report("salvage-system-announcement-lost");
+                    magnet.MagnetState = new MagnetState(MagnetStateType.CoolingDown, currentTime + CooldownTime);
+                    break;
+                case MagnetStateType.CoolingDown:
+                    magnet.MagnetState = MagnetState.Inactive;
+                    break;
+            }
+        }
+
+        public override void Update(float frameTime)
+        {
+            foreach (var gridIdAndState in _salvageGridStates)
+            {
+                var state = gridIdAndState.Value;
+                if (state.ActiveMagnets.Count == 0) continue;
+                var gridId = gridIdAndState.Key;
+                // Not handling the case where the salvage we spawned got paused
+                // They both need to be paused, or it doesn't make sense
+                if (_pauseManager.IsGridPaused(gridId)) continue;
+                state.CurrentTime += TimeSpan.FromSeconds(frameTime);
+                var currentMagnet = 0;
+                while (currentMagnet < state.ActiveMagnets.Count)
+                {
+                    var magnet = state.ActiveMagnets[currentMagnet];
+                    if (magnet.MagnetState.Until > state.CurrentTime)
+                    {
+                        currentMagnet++;
+                        continue;
+                    }
+                    Transition(magnet, state.CurrentTime);
+                    if (magnet.MagnetState.StateType == MagnetStateType.Inactive)
+                    {
+                        state.ActiveMagnets.RemoveAt(currentMagnet);
+                    }
+                    else
+                    {
+                        currentMagnet++;
+                    }
+                }
```

Comment:
Can't you just use a RemQueue / iterate in reverse order?

Reply:
Switched to using a RemQueue.